src/Compilers/Core/Portable/DiagnosticAnalyzer/SourceOrNonSourceFile.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Represents a source file or a non-source file.
+    /// For source files, <see cref="SourceTree"/> is non-null and <see cref="NonSourceFile"/> is null.
+    /// For non-source files, <see cref="NonSourceFile"/> is non-null and <see cref="SourceTree"/> is null.
+    /// </summary>
+    internal abstract class SourceOrNonSourceFile
+        : IEquatable<SourceOrNonSourceFile>
```

Comment:
Consider implementing `operator==` and company given that this is an equatable type